tgui/packages/tgui/interfaces/NtosMessenger/ChatScreen.tsx

```diff
@@ -0,0 +1,407 @@
+  sendingVirus: BooleanLike;
+};
+
+type ChatScreenState = {
+  msg: string;
+  selectingPhoto: boolean;
+  canSend: boolean;
+  previewingImage: string | null;
+};
+
+const READ_UNREADS_TIME_MS = 3000;
+const SEND_COOLDOWN_MS = 1000;
+
+export class ChatScreen extends Component<ChatScreenProps, ChatScreenState> {
+  scrollRef: RefObject<HTMLDivElement>;
+  readUnreadsTimeout: NodeJS.Timeout | null;
+
+  state: ChatScreenState = {
+    msg: '',
+    selectingPhoto: false,
+    canSend: true,
+    previewingImage: null,
+  };
+
+  constructor(props: ChatScreenProps) {
+    super(props);
+
+    this.scrollRef = createRef();
+
+    this.scrollToBottom = this.scrollToBottom.bind(this);
+    this.handleMessageInput = this.handleMessageInput.bind(this);
+    this.handleSelectPicture = this.handleSelectPicture.bind(this);
+    this.handleSendMessage = this.handleSendMessage.bind(this);
+    this.trySetReadTimeout = this.trySetReadTimeout.bind(this);
+    this.tryClearReadTimeout = this.tryClearReadTimeout.bind(this);
+    this.clearUnreads = this.clearUnreads.bind(this);
+  }
+
+  componentDidMount() {
+    this.scrollToBottom();
+    this.trySetReadTimeout();
+  }
+
+  componentDidUpdate(
+    prevProps: ChatScreenProps,
+    _prevState: ChatScreenState,
+    _snapshot: any
+  ) {
+    if (prevProps.chat?.messages.length !== this.props.chat?.messages.length) {
+      this.scrollToBottom();
+      this.trySetReadTimeout();
+    }
+  }
+
+  componentWillUnmount() {
+    const { act } = useBackend(this.context);
+
+    const chat = this.props.chat;
+    if (!chat) return;
+
+    this.tryClearReadTimeout();
+
+    act('PDA_saveMessageDraft', {
+      ref: chat.ref,
+      msg: this.state.msg,
+    });
+  }
+
+  trySetReadTimeout() {
+    const chat = this.props.chat;
+
+    this.tryClearReadTimeout();
+
+    if (chat && chat.unread_messages > 0) {
+      this.readUnreadsTimeout = setTimeout(() => {
+        this.clearUnreads();
+        this.readUnreadsTimeout = null;
+      }, READ_UNREADS_TIME_MS);
+    }
+  }
+
+  tryClearReadTimeout() {
+    if (this.readUnreadsTimeout) {
+      clearTimeout(this.readUnreadsTimeout);
+      this.readUnreadsTimeout = null;
+    }
+  }
+
+  clearUnreads() {
+    const { act } = useBackend(this.context);
+    const chat = this.props.chat;
+    act('PDA_clearUnreads', { ref: chat!.ref });
+  }
+
+  scrollToBottom() {
+    const scroll = this.scrollRef.current;
+    if (scroll !== null) {
+      scroll.scrollTop = scroll.scrollHeight;
+    }
+  }
+
+  handleSelectPicture() {
+    const { isSilicon } = this.props;
+    const { act } = useBackend(this.context);
+    if (isSilicon) {
+      act('PDA_siliconSelectPhoto');
+    } else {
+      this.setState({ selectingPhoto: true });
+    }
+  }
+
+  handleSendMessage() {
+    if (this.state.msg === '') return;
+    const { act } = useBackend(this.context);
+    const { temp_msgr, chat } = this.props;
+
+    let ref = chat ? chat.ref : temp_msgr!.ref!;
```

Comment:
Okay, the way you do this normally is somehting like this:

```ts
type MyType = {
    // everything that is always enabled
} & (
    { chat: Chat } | { tempMessenger: Messenger }
)
```

It's an intersection of the type with a union, meaning one or the other exists.

You then check `"chat" in props` in order to narrow it.